cmd/routed-eni-cni-plugin/cni.go

```diff
@@ -325,13 +322,16 @@ func del(args *skel.CmdArgs, cniTypes typeswrapper.CNITYPES, grpcClient grpcwrap
 			args.ContainerID, err)
 
 		// When IPAMD is unreachable, try to teardown pod network using previous result. This action prevents rules from leaking while IPAMD is unreachable.
-		// Note that connection error is still returned to kubelet so that delete retries until IPAMD can be notified.
+		// Note that no error is returned to kubelet as there is no guarantee that kubelet will retry delete, and returning an error would prevent container runtime
+		// from cleaning up resources. When IPAMD again becomes responsive, it is responsible for reclaiming IP.
```

Comment:
Say if kubelet doesn't retry and if IPAMD is recovering from the state file then reclaiming IP is not possible right? Am is missing anything here?

Reply:
Ideally I feel, IPAMD cleanup should happen at the end, i.e CNI should go ahead and cleanup resources (teardownnetwork) in the networking stack and then invoke IPAMD (delnetwork) to release IP.. that will also help with leaked rules..we can discuss offline more..

Reply:
IPAMD will reclaim the IP from the ENI because it will detect that its entry in the state file is invalid. CNI does clean up all of the resources (really just the IP rules) if the pod was created by v1.12.1+. It is only if there is no PrevResult that the IP rules could leak and so IPAMD has to try to clear them

Reply:
TBH, ideally i think we don't need this change, it should be kubelet's job to make sure always retry call cniDel even the Pod have been gone from APIServer. Kubelet should fix their bug. and this change is just our workaround for this kubelet flaw.

also, let's test the logic that IPAMD can reclaim the IP in this case.

Reply:
From kubelet perspective, this is working as expected: the pod has disappeared from the API server, so there is nothing to retry. In the case that motivated this PR, it is because the pod was force deleted, but I don't see that as a kubelet bug.

Outside of kubelet, though, the scheduling deadlock still exists, where the aws-node pod cannot be placed until pods are deleted, and pods cannot be deleted unless IPAMD is running.

Reply:
I did confirm from testing that IPAMD reclaims the IP from the ENI and can reuse it later